### PR TITLE
DRILL-7746: Add REST API Ability to Accept CSV Input

### PR DESCRIPTION
**Pull request by cgivre**
# [DRILL-7746](https://issues.apache.org/jira/browse/DRILL-7746): Add REST API Ability to Accept CSV Input

## Description
This test enables the REST plugin to query APIs that return CSV as well as JSON. 
## Documentation
This adds an additional configuration option for each connection: `inputType` which defaults to JSON. If the API outputs `csv` then user should set `inputType` to `csv`.  The plugin requires that the data be in CSV format with the first line being the column headers.  

This PR uses the same library as Drill currently does for CSV parsing, however it was not possible to use the `CompliantTextReader` without extensive modifications to that file.   This PR implements a simple CSV reader using the same underlying parser.

## Testing
Added unit test for CSV input.  Ran tests from CLI as well.

### CODE REVIEW

File: contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/HttpScanBatchCreator.java [hunk 1]
```diff
@@ -92,9 +92,16 @@ public void bind(ManagedScanFramework framework) { }
...
+      String inputType = subScan.tableSpec().connectionConfig().inputType();
+
       // Only a single scan (in a single thread)
       if (count++ == 0) {
-        return new HttpBatchReader(subScan);
+        if (inputType.equalsIgnoreCase("csv")) {
+          return new HttpCSVBatchReader(subScan);
+        } else {
+          return new HttpBatchReader(subScan);
+        }
```
Comment (sanel) [inline contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/HttpScanBatchCreator.java:104]: Github won't let me leave the comment for `else` block down below, but that `return null;` can go as default (without `else`).

File: contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/HttpCSVBatchReader.java [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+    return builder.buildSchema();
+  }
+
+  private void populateWriterArray() {
+    columnWriters = new ArrayList<>();
+
+    // Case for empty result set
+    if (firstRow == null || firstRow.length == 0) {
+      return;
+    }
```
Comment (sanel) [inline contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/HttpCSVBatchReader.java:133]: Can `columnWriters = new ArrayList<>();` go here (after `firstRow` checks)?

File: contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/HttpCSVBatchReader.java [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+    return true;
+  }
+
+  private TupleMetadata buildSchema() {
+
+    this.firstRow = csvReader.parseNext();
+
+    // Case for empty dataset
+    if (firstRow == null) {
```
Comment (sanel) [inline contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/HttpCSVBatchReader.java:115]: Does it make sense to refactor complete function to this:

```java
this.firstRow = csvReader.parseNext();
if (firstRow != null) {
  for (String value : firstRow) {
      builder.addNullable(value, TypeProtos.MinorType.VARCHAR);
    }
}
return builder.buildSchema();
```

File: contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/HttpCSVBatchReader.java [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+    // CSV loader setup
+    InputStream inStream = http.getInputStream();
+
+    this.csvReader = new CsvParser(csvSettings);
+    csvReader.beginParsing(inStream);
+
+    // Build the Schema
+    builder = new SchemaBuilder();
+    TupleMetadata drillSchema = buildSchema();
```
Comment (sanel) [inline contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/HttpCSVBatchReader.java:89]: If `buildSchema()` is called only once, I presume `csvReader.beginParsing(inStream)` can go in it, to localize state; something like:

```java
buildSchema(http.getInputStream());
```

File: contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/HttpScanBatchCreator.java (hunk 1 above)
Comment (cgivre) [reply contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/HttpScanBatchCreator.java:104]: Fixed

File: contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/HttpCSVBatchReader.java (hunk 2 above)
Comment (cgivre) [reply contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/HttpCSVBatchReader.java:131]: Sure can!  Fixed.

File: contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/HttpCSVBatchReader.java (hunk 3 above)
Comment (cgivre) [reply contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/HttpCSVBatchReader.java]: Indeed... Much cleaner.  Fixed.

File: contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/HttpCSVBatchReader.java (hunk 4 above)
Comment (cgivre) [reply contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/HttpCSVBatchReader.java:91]: I realized from this comment that I didn't implement the `close()` method.  While `buildSchema()` is only called once, the proper way to implement these readers is to explicitly close any inputstreams, readers or whatever.  I fixed that.

**Comment by cgivre** [comment]
@sanel 
Thank you very much for the quick review!  I addressed your comments.  Please let me know if we're good to go.  If so I'll squash commits and rebase tomorrow.

**Comment by sanel** [comment]
Great! Looks good.

**Comment by cgivre** [comment]
Thanks @sanel! 
Commits squashed!

**Pull request merged**